storage/storage_image.go

```diff
@@ -476,14 +550,175 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
+		}
+		if err = s.imageRef.transport.store.SetNames(imageID, names); err != nil {
+			logrus.Debugf("error setting names %v on image %q: %v", names, imageID, err)
+			return errors.Wrapf(err, "error setting names %v on image %q", names, imageID)
+		}
+		logrus.Debugf("set names of image %q to %v", imageID, names)
+	}
+	return nil
+}
+
+func (s *storageImageDestination) commitGroup(instanceDigest *digest.Digest) error {
+	// Find the list of non-layer blobs.
+	instance := instanceFor(instanceDigest)
+	m := s.rawManifests[instance]
+	if len(m) == 0 {
 		return errors.New("Internal error: storageImageDestination.Commit() called without PutManifest()")
 	}
-	man, err := manifest.FromBlob(s.manifest, manifest.GuessMIMEType(s.manifest))
+	man, err := manifest.ListFromBlob(m, s.manifestTypes[instance])
+	if err != nil {
+		return errors.Wrapf(err, "Internal error: storageImageDestination.Commit() called without PutManifest()")
+	}
+	// Compute the manifest digest.
+	options := &storage.ImageOptions{}
+	if manifestDigest, err := manifest.Digest(m); err == nil {
+		options.Digest = manifestDigest
+	}
+	// Create the image record, pointing to no layers.
+	intendedID := s.imageRef.id
+	if instanceDigest != nil {
+		intendedID = ""
+	}
+	if intendedID == "" {
+		intendedID = man.ImageID()
+	}
+	return s.commitImageRecord(instanceDigest, m, intendedID, "", options)
+}
+
+func (s *storageImageDestination) commitImageRecord(instanceDigest *digest.Digest, manifest []byte, intendedID string, layer string, options *storage.ImageOptions) error {
+	oldNames := []string{}
+	img, err := s.imageRef.transport.store.CreateImage(intendedID, nil, layer, "", options)
 	if err != nil {
-		return errors.Wrapf(err, "error parsing manifest")
+		if errors.Cause(err) != storage.ErrDuplicateID {
+			logrus.Debugf("error creating image: %q", err)
+			return errors.Wrapf(err, "error creating image %q", intendedID)
+		}
+		img, err = s.imageRef.transport.store.Image(intendedID)
+		if err != nil {
+			return errors.Wrapf(err, "error reading image %q", intendedID)
+		}
+		if img.TopLayer != layer {
+			why := "uses layers"
+			if layer != "" {
+				why = "uses a different top layer"
+			}
+			logrus.Debugf("error creating image: image with ID %q exists, but %s", intendedID, why)
+			return errors.Wrapf(storage.ErrDuplicateID, "image with ID %q already exists, but %s", intendedID, why)
+		}
+		logrus.Debugf("reusing image ID %q", img.ID)
+		oldNames = append(oldNames, img.Names...)
+	} else {
+		logrus.Debugf("created new image ID %q", img.ID)
+	}
+	// Add the non-layer blobs as data items.  Since we only share layers, they should all be in files, so
+	// we just need to screen out the ones that are actually layers to get the list of non-layers.
+	if err = s.commitDataBlobs(img.ID); err != nil {
+		if _, err2 := s.imageRef.transport.store.DeleteImage(img.ID, true); err2 != nil {
+			logrus.Debugf("error deleting incomplete image %q: %v", img.ID, err2)
```

Comment:
On the fence here, but if you've an error deleting an incomplete image, should it just go to debug  or should you wrap up err2 also and return with the return below?

Reply:
Ditto with the next few delete with a debug on error.

Reply:
`copy.Image` does that in a few (very few) cases, see `if err := dest.Close(); err != nil {`. All too often we do `defer something.Close()` and ignore the possible failure.

Either way it’s problematic. On one hand, yes, throwing away failure information may make diagnosing failures difficult. On the other, the `copy.Image` attempt to parenthesize the secondary failure has already confused at least one user, who focused on the secondary failure instead of the primary one.